week-1-ts-ecosystem/w01e02-enter-the-world-of-typescript/src/framework/dom-creators.ts

```diff
@@ -1,12 +1,12 @@
 function createElement(tagName = 'div', className = '') {
-  const element = document.createElement(tagName)
+  const element = document.createElement(tagName);
   if (className.length) {
-    element.className = className
+    element.className = className;
   }
-  return element
+  return element;
 }
 
-export const p = (className = '') => createElement('p', className)
-export const div = (className = '') => createElement('div', className)
-export const section = (className = '') => createElement('section', className)
-export const article = (className = '') => createElement('article', className)
+export const p = (className = '') => createElement('p', className);
```

Comment:
Tutaj ładnie by było dodać / poprawić typowanie.

Czy byłbyś w stanie zrobić tak żeby _TS_ podpowiadał że faktycznie `p` zwraca `HTMLParagraphElement` a `div` to `HTMLDivElement` a pozostałe `HTMLElement` ?

Reply:
To jest ciężkie. Można zrobić mapowanie typów tutaj. HTML generalnie jest luźny jeżeli chodzi o tagi. W większości pozwala na obojętnie jaki string. Nie mam pomysłu jak to rozwiązać można na sztywno mapować rezultat ale chyba nie o to Ci chodzi w tym zadaniu ?

Reply:
```ts
function createElement<K extends keyof HTMLElementTagNameMap>(tagName = 'div', className = ''): HTMLElementTagNameMap[K] {
  const element = document.createElement(tagName) as HTMLElementTagNameMap[K];
  if (className.length) {
    element.className = className;
  }
  return element;
}


export const p = (className = ''): HTMLParagraphElement | null => createElement<'p'>('p', className);
export const div = (className = '') => createElement<'div'>('div', className);
export const section = (className = '') => createElement<'section'>('section', className);
export const article = (className = '') => createElement<'article'>('article', className);

`

Reply:
@michaljabi